### PR TITLE
Lots of keybindings for emacs defaults: rectangles, kmacro, timeclock, insert-buffer.

### PR DESCRIPTION
**Pull request by equwal**
All these bindings are for functions which are in emacs by default, and which I
think deserve keybindings. I have been using these as personal layers for
awhile, but I think others might benefit from them.

Changes: Add ediff binding
 Add rectangle [ keybinding
 Add kmacro bindings
 Add timeclock keybindings
 Add insert-buffer keybinding -- slurp another buffer in this one

### CODE REVIEW

File: layers/+spacemacs/spacemacs-defaults/keybindings.el [hunk 1]
```diff
@@ -134,6 +138,54 @@
...
 ;; shell command  -------------------------------------------------------------
 (spacemacs/set-leader-keys "!" 'shell-command)
+;; kmacros --------------------------------------------------------------------
+(spacemacs/set-leader-keys
+  "km2c" 'kmacro-call-ring-2nd
+  "km2C" 'kmacro-call-ring-2nd-repeat
+  "km2v" 'kmacro-view-ring-2nd
+  "kmb" 'kmacro-bind-to-key
+  "kmc" 'kmacro-add-counter
+  "kmc" 'kmacro-call-macro
```
[the remark this answers is not in the document]
Comment (equwal) [reply layers/+spacemacs/spacemacs-defaults/keybindings.el]: I renamed them and updated the entry.

File: layers/+spacemacs/spacemacs-defaults/keybindings.el [hunk 2]
```diff
@@ -134,6 +138,54 @@
...
+  "kmls" 'kmacro-edit-lossage
+  "kmm" 'kmacro-end-call-mouse
+  "kmn" 'kmacro-cycle-ring-next
+  "kmN" 'kmacro-cycle-ring-previous
+  "kmp" 'kmacro-cycle-ring-previous
+  "kmr" 'helm-register
+  "kmsc" 'kmacro-set-counter
+  "kmsf" 'kmacro-set-format
+  "kmsr" 'kmacro-swap-ring
+  "kmte" 'kmacro-step-edit-macro
```
[the remark this answers is not in the document]
Comment (equwal) [reply layers/+spacemacs/spacemacs-defaults/keybindings.el]: Indeed. Thanks for the feedback.

**Comment by equwal** [comment]
> When restarting Spacemacs, before opening a `emacs-lisp-mode` buffer.
> `SPC k` only shows: `m -> kmacro`
> 
> But after opening for example `.spacemacs`, then lots of `evil-lisp-state` key bindings are shown under the `SPC k` prefix, but now `m -> kmacro` isn't shown anymore and now
> `SPC k m` returns: `SPC k m is undefined`
> 
> There's a typo in the commit message:
> Before: `ediff: T;`
> After: `ediff: E;`
> 
> And these key binding prefixes are missing:
> 
> ```emacs-lisp
>                                        ("Eb"  "buffers")
>                                        ("Ed"  "directories")
>                                        ("Ef"  "files")
>                                        ("Em"  "merge")
>                                        ("Emb" "buffers")
>                                        ("Emd" "directories")
>                                        ("Emf" "files")
>                                        ("Emr" "revisions")
>                                        ("Er"  "regions")
>                                        ("Ew"  "windows")
> ```
> 
> and
> 
> ```emacs-lisp
>                                        ("km2" "ring")
>                                        ("kme" "edit")
>                                        ("kml" "name/edit lossage")
>                                        ("kms" "set/swap")
> ```
> 
> I don't know if there's a better name for `name/edit lossage`, they are the only keys under the
> `SPC k m l` prefix:
> 
> ```
>   "kmlm"   'kmacro-name-last-macro
>   "kmls"   'kmacro-edit-lossage
> ```

I added those prefixes and dumped that prefix. I don't yet know the cause of the bug where `SPC k` is shadowed by `SPC k m`.

**Comment by equwal** [comment]
I don't think this is my bug. A quick `grep -r backward-delete` shows that these keys are not bound anywhere in the develop branch.

**Comment by equwal** [comment]
> When restarting Spacemacs, before opening a `emacs-lisp-mode` buffer.
> `SPC k` only shows: `m -> kmacro`
> 
> But after opening for example `.spacemacs`, then lots of `evil-lisp-state` key bindings are shown under the `SPC k` prefix, but now `m -> kmacro` isn't shown anymore and now
> `SPC k m` returns: `SPC k m is undefined`

This issue is not my making. See this issue: https://github.com/syl20bnr/spacemacs/issues/12963

**Comment by duianto** [comment]
Maybe the kmacro key bindings could be added to `SPC K` or `SPC M` since `SPC k` seems to replace any previous bindings under that prefix with the `evil-lisp-state-map` when a `prog-mode` buffer is opened.

**Comment by equwal** [comment]
> Maybe the kmacro key bindings could be added to `SPC K` or `SPC M` since `SPC k` seems to replace any previous bindings under that prefix with the `evil-lisp-state-map` when a `prog-mode` buffer is opened.

Okay, I think `SPC K` is a nicer keymap anyway.

**Comment by equwal** [comment]
This does not help this other issue: https://github.com/syl20bnr/spacemacs/issues/12963

**Comment by duianto** [comment]
Nope, it just binds this PRs suggested kmacro keys under the `SPC K` (uppercase) prefix.

**Comment by equwal** [comment]
Thank you for your edits. I went through and tested the functions, finding two erroneous ones. Let me know if you find anything else that needs done.

**Comment by duianto** [comment]
```
  "Ec"   'ediff-backup
```
Instead of `SPC E c` for backup, `SPC E B` might be a better mnemonic.

---

```
  "Ken" 'kmacro-name-last-macro
```
This is under the `edit` prefix.
If it can be considered as editing a macro, then it's probably fine.
Otherwise it might be more visible outside the edit prefix. But I'm not sure where.

---

```
  "Ed."   'spacemacs/ediff-dotfile-and-template
```
`spacemacs/ediff-dotfile-and-template` compares two files but it's under the directories prefix. Move it under the files prefix: `SPC E f .`

---

```
"K<f4>" 'kmacro-end-or-call-macro-repeat
```
is accessed with: `SPC K < f 4 >` (typed as individual keys).
It should probably have a non function key binding, because `F4` calls the same command without any prefix keys by default.

And `kmacro-start-macro-or-insert-counter` which is called with `F3` by default, would probably also be useful as a `SPC K` key binding. Can't think of where though.

---

```
  "Rc" 'close-rectangle
  "Rc" 'rectangle-number-lines
```
Duplicate key bindings. Only the last one is defined.

**Comment by equwal** [comment]
> ```
>   "Ec"   'ediff-backup
> ```
> 
> Instead of `SPC E c` for backup, `SPC E B` might be a better mnemonic.

- [X]

> 
> ```
>   "Ken" 'kmacro-name-last-macro
> ```
> 
> This is under the `edit` prefix.
> If it can be considered as editing a macro, then it's probably fine.
> Otherwise it might be more visible outside the edit prefix. But I'm not sure where.

- [X] Just use `SPC K N`, no need to duplicate Kp.

> 
> ```
>   "Ed."   'spacemacs/ediff-dotfile-and-template
> ```
> 
> `spacemacs/ediff-dotfile-and-template` compares two files but it's under the directories prefix. Move it under the files prefix: `SPC E f .`

- [X] OK.
> 
> ```
> "K<f4>" 'kmacro-end-or-call-macro-repeat
> ```
> 
> is accessed with: `SPC K < f 4 >` (typed as individual keys).
> It should probably have a non function key binding, because `F4` calls the same command without any prefix keys by default.
> 

Oops. Looks like I can only use function keys if they are defined as prefixes,
unless I do not use that macro.
- [X] Move to `SPC K )`, which has the benefit of 1) being like Emacs and 2) being at the top of `which-key`.

> And `kmacro-start-macro-or-insert-counter` which is called with `F3` by default, would probably also be useful as a `SPC K` key binding. Can't think of where though.
> 

- [X] How about `SPC K (`

> ```
>   "Rc" 'close-rectangle
>   "Rc" 'rectangle-number-lines
> ```
> 
> Duplicate key bindings. Only the last one is defined.

- [X] Use `SPC R N` similar to above (no need to duplicate `SPC R p`).

**Comment by duianto** [comment]
Thank you for contributing to Spacemacs. It's safe to delete your branch.
The PRs changes have been cherry picked into the `develop` branch.
Commit(s): https://github.com/syl20bnr/spacemacs/commit/cab1e69585f7ad47613e45d7123ebfe64bba0a7e

I missed that there was a typo in the commit subject kmacro key binding, it's supposed to say: `K`

**Comment by duianto** [comment]
And we need to document the key bindings.

**Comment by duianto** [comment]
`SPC R t` which calls `transpose-regions` shows the message:
>command-execute: Wrong type argument: commandp, transpose-regions

Meaning that it isn't a command that can be called from either a key binding or `M-x`

It expects four arguments:
>(transpose-regions STARTR1 ENDR1 STARTR2 ENDR2 &optional LEAVE-MARKERS)
> 
> Transpose region STARTR1 to ENDR1 with STARTR2 to ENDR2.
> The regions should not be overlapping, because the size of the buffer is
> never changed in a transposition.
> 
> Optional fifth arg LEAVE-MARKERS, if non-nil, means don’t update
> any markers that happen to be located in the regions.
> 
> Transposing beyond buffer boundaries is an error.
>

**Comment by syl20bnr** [comment]
@equwal I moved the key bindings for ediff to `SPC D` for `Diff` instead. See commit https://github.com/syl20bnr/spacemacs/commit/51c7fb961
I also added documentation, feel free to review it and correct the description as I'm not an expert with ediff :-)

Note that I will likely move other key bindings of this PR in the future.

**Comment by syl20bnr** [comment]
@equwal I updated the key bindings for keyboard macros in an attempt to make them self documented/usable using only which-key.

**Comment by equwal** [comment]
> @equwal I updated the key bindings for keyboard macros in an attempt to make them self documented/usable using only which-key.

Great! I hope that fixes this bug mentioned by @duianto.

> But after opening for example `.spacemacs`, then lots of `evil-lisp-state` key bindings are shown under the `SPC k` prefix, but now `m -> kmacro` isn't shown anymore and now
> `SPC k m` returns: `SPC k m is undefined`

**Comment by duianto** [comment]
Yes, that's not an issue anymore, now that the macro key bindings are under the `SPC K` prefix.

**Pull request closed without merging**